File: src/git.ts
export type Entry = {
  indexStatus: string;
  worktreeStatus: string;
  path: string;
  renamedFrom?: string;
};

const EDITABLE_CODES = new Set(["M", "A", "D", "R", "C", "T", "U", "?"]);

export function filterEditable(entries: Entry[]): Entry[] {
  return entries.filter(
    (e) =>
      EDITABLE_CODES.has(e.indexStatus) ||
      EDITABLE_CODES.has(e.worktreeStatus),
  );
}

export async function findRepoRoot(cwd: string = Deno.cwd()): Promise<string> {
  const cmd = new Deno.Command("git", {
    args: ["rev-parse", "--show-toplevel"],
    cwd,
    stdout: "piped",
    stderr: "piped",
  });
  const { code, stdout, stderr } = await cmd.output();
  if (code !== 0) {
    throw new Error(
      `not a git repository: ${new TextDecoder().decode(stderr).trim()}`,
    );
  }
  return new TextDecoder().decode(stdout).trimEnd();
}

export async function listChanges(repoRoot: string): Promise<Entry[]> {
  const cmd = new Deno.Command("git", {
    args: ["status", "--porcelain=v1", "-z"],
    cwd: repoRoot,
    stdout: "piped",
    stderr: "piped",
  });
  const { code, stdout, stderr } = await cmd.output();
  if (code !== 0) {
    throw new Error(
      `git status failed: ${new TextDecoder().decode(stderr).trim()}`,
    );
  }
  return filterEditable(parsePorcelain(stdout));
}

export function parsePorcelain(bytes: Uint8Array): Entry[] {
  const text = new TextDecoder().decode(bytes);
  if (text.length === 0) return [];

  const records = text.split("\0");
  if (records.at(-1) === "") records.pop();

  const entries: Entry[] = [];
  let i = 0;
  while (i < records.length) {
    const rec = records[i];
    if (rec.length < 3) {
      i++;
      continue;
    }
    const indexStatus = rec[0];
    const worktreeStatus = rec[1];
    const path = rec.slice(3);
    const isRenameOrCopy = indexStatus === "R" || indexStatus === "C" ||
      worktreeStatus === "R" || worktreeStatus === "C";
    if (isRenameOrCopy && i + 1 < records.length) {
      entries.push({
        indexStatus,
        worktreeStatus,
        path,
        renamedFrom: records[i + 1],
      });
      i += 2;
    } else {
      entries.push({ indexStatus, worktreeStatus, path });
      i += 1;
    }
  }
  return entries;
}

File: git.go
package main

import "bytes"

type Entry struct {
	IndexStatus    byte
	WorktreeStatus byte
	Path           string
	RenamedFrom    string
}

var editableCodes = map[byte]bool{
	'M': true, 'A': true, 'D': true, 'R': true,
	'C': true, 'T': true, 'U': true, '?': true,
}

func filterEditable(entries []Entry) []Entry {
	out := make([]Entry, 0, len(entries))
	for _, e := range entries {
		if editableCodes[e.IndexStatus] || editableCodes[e.WorktreeStatus] {
			out = append(out, e)
		}
	}
	return out
}

func parsePorcelain(data []byte) []Entry {
	if len(data) == 0 {
		return nil
	}
	records := bytes.Split(data, []byte{0})
	if len(records) > 0 && len(records[len(records)-1]) == 0 {
		records = records[:len(records)-1]
	}
	var out []Entry
	for i := 0; i < len(records); {
		rec := records[i]
		if len(rec) < 3 {
			i++
			continue
		}
		idx := rec[0]
		wt := rec[1]
		path := string(rec[3:])
		renameOrCopy := idx == 'R' || idx == 'C' || wt == 'R' || wt == 'C'
		if renameOrCopy && i+1 < len(records) {
			out = append(out, Entry{
				IndexStatus:    idx,
				WorktreeStatus: wt,
				Path:           path,
				RenamedFrom:    string(records[i+1]),
			})
			i += 2
		} else {
			out = append(out, Entry{
				IndexStatus:    idx,
				WorktreeStatus: wt,
				Path:           path,
			})
			i++
		}
	}
	return out
}
